sway-ir/src/optimize/cse.rs

```diff
@@ -0,0 +1,408 @@
+    match &instr.get_instruction(context).unwrap().op {
+        InstOp::AsmBlock(_, _) => None,
+        InstOp::UnaryOp { op, arg } => Some(Expr::UnaryOp {
+            op: *op,
+            arg: vntable.value_map.get(arg).cloned().unwrap(),
+        }),
+        InstOp::BinaryOp { op, arg1, arg2 } => Some(Expr::BinaryOp {
+            op: *op,
+            arg1: vntable.value_map.get(arg1).cloned().unwrap(),
+            arg2: vntable.value_map.get(arg2).cloned().unwrap(),
+        }),
+        InstOp::BitCast(val, ty) => Some(Expr::BitCast(
+            vntable.value_map.get(val).cloned().unwrap(),
+            *ty,
+        )),
+        InstOp::Branch(_) => None,
+        InstOp::Call(_, _) => None,
+        InstOp::CastPtr(val, ty) => Some(Expr::CastPtr(
+            vntable.value_map.get(val).cloned().unwrap(),
+            *ty,
+        )),
+        InstOp::Cmp(pred, val1, val2) => Some(Expr::Cmp(
+            *pred,
+            vntable.value_map.get(val1).cloned().unwrap(),
+            vntable.value_map.get(val2).cloned().unwrap(),
+        )),
+        InstOp::ConditionalBranch { .. } => None,
+        InstOp::ContractCall { .. } => None,
+        InstOp::FuelVm(_) => None,
+        InstOp::GetLocal(_) => None,
+        InstOp::GetConfig(_, _) => None,
+        InstOp::GetElemPtr {
+            base,
+            elem_ptr_ty,
+            indices,
+        } => Some(Expr::GetElemPtr {
+            base: vntable.value_map.get(base).cloned().unwrap(),
+            elem_ptr_ty: *elem_ptr_ty,
+            indices: indices
+                .iter()
+                .map(|idx| vntable.value_map.get(idx).cloned().unwrap())
+                .collect(),
+        }),
+        InstOp::IntToPtr(val, ty) => Some(Expr::IntToPtr(
+            vntable.value_map.get(val).cloned().unwrap(),
+            *ty,
+        )),
+        InstOp::Load(_) => None,
+        InstOp::MemCopyBytes { .. } => None,
+        InstOp::MemCopyVal { .. } => None,
+        InstOp::Nop => None,
+        InstOp::PtrToInt(val, ty) => Some(Expr::PtrToInt(
+            vntable.value_map.get(val).cloned().unwrap(),
+            *ty,
+        )),
+        InstOp::Ret(_, _) => None,
+        InstOp::Store { .. } => None,
+    }
+}
+
+/// Convert a PHI argument to Expr
+fn phi_to_expr(context: &Context, vntable: &VNTable, phi_arg: Value) -> Expr {
+    let phi_arg = phi_arg.get_argument(context).unwrap();
+    let phi_args = phi_arg
+        .block
+        .pred_iter(context)
+        .map(|pred| {
+            let incoming_val = phi_arg
+                .get_val_coming_from(context, pred)
+                .expect("No parameter from predecessor");
+            vntable.value_map.get(&incoming_val).cloned().unwrap()
+        })
+        .collect();
+    Expr::Phi(phi_args)
+}
+
+#[derive(Default)]
+struct VNTable {
+    value_map: FxHashMap<Value, ValueNumber>,
+    expr_map: FxHashMap<Expr, ValueNumber>,
+}
+
+impl Debug for VNTable {
```

Comment:
Makes sense to print this to the ir tests? I think it would be interesting to see.

Reply:
I gave a quick try to see if we can use the `tracing` crate (which we already use in other parts of `forc`), but I think (and as you suggested offline) we should do a separate PR to get that working and print such information for most other analyses / optimizations.